x/bank/internal/keeper/keeper.go

```diff
@@ -380,8 +381,8 @@ func (keeper BaseViewKeeper) Codespace() sdk.CodespaceType {
 }
 
 // CONTRACT: assumes that amt is valid.
-func trackDelegation(acc exported.Account, blockTime time.Time, amt sdk.Coins) error {
-	vacc, ok := acc.(exported.VestingAccount)
+func trackDelegation(acc authexported.Account, blockTime time.Time, amt sdk.Coins) error {
```

Comment:
We need a larger discussion on how to handle this. Ideally, bank would not rely on vesting somehow. Thoughts @AdityaSripal?

Reply:
I think it’s time to move these 2 funcs back to staking

Reply:
`(Delegate|Undelegate)Coins` should be removed from `x/bank` Keeper too along with the supply related methods. I'm not sure of the best way to do this. Maybe function types that can be set on the keeper.

Reply:
I don't see a great way to do this either. If we move `(Delegate|Undelegate)Coins` from `x/bank` to `x/supply` for example, then `x/supply` must rely on vesting.

I think any delegation logic has to be aware of the existence of VestingAccounts as a special case. One option could be bringing this responsibility to all Accounts:

So add `DelegateCoins`/`UndelegateCoins` to the `Account` interface and have `x/supply` call this function when it wants to perform delegation functions. The problem there is adding some pretty high-level logic to the `Account` interface, which is not ideal.

I think the best way to tackle this is a separate PR, where you can allow accounts to register some hooks to run upon some action (`send`/`delegate`)